### PR TITLE
Archive setup

### PR DESCRIPTION
**Pull request by tongzhi1998**
Monad Archive Setup
To run an archive instance:

`CXX=/usr/bin/g++-13 CC=/usr/bin/gcc-13 ASMFLAGS=-march=haswell CFLAGS="-march=haswell" CXXFLAGS="-march=haswell" TRIEDB_TARGET=triedb_driver cargo run --package monad-archive -- --s3-bucket <> --triedb-path <>`

Benchmark Stats:
Archiving 100 blocks from 12M: 8s
Archiving 500 blocks from 12M: 43s

Testing branch: `https://github.com/monad-crypto/monad-bft/tree/tzhi/archive_interface_testing`

### CODE REVIEW

File: monad-archive/src/archive_interface.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    // eth_getBlockByNumber
+    // eth_getRawBlock
+    // eth_getRawHeader
+    async fn get_block_by_number(&self, block_num: u64) -> Result<Block, ArchiveError>;
+
+    //eth_getBlockTransactionCountByHash
+    async fn get_block_transaction_count_by_hash(
+        &self,
+        block_hash: &[u8],
```
Comment (eerkaijun) [inline monad-archive/src/archive_interface.rs]: would recommend setting this as &[u8; 32] for stricter typings

File: monad-archive/src/archive_interface.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    // eth_getTransactionByBlockNumberAndIndex
+    async fn get_transaction_by_block_number_and_index(
+        &self,
+        block_number: u64,
+        tx_index: u64,
+    ) -> Result<Transaction, ArchiveError>;
+
+    // eth_getTransactionByHash
+    // eth_getRawTransaction
+    async fn get_transaction_by_hash(&self, tx_hash: &[u8]) -> Result<Transaction, ArchiveError>;
```
Comment (eerkaijun) [inline monad-archive/src/archive_interface.rs]: same as above

**Comment by eerkaijun** [comment]
are there any rough performance intuition on uploading data and reading data from archive? e.g. if the load of the chain is 1k TPS, can we also read 1k data from the archive?

**Comment by tongzhi1998** [comment]
> reading



> are there any rough performance intuition on uploading data and reading data from archive? e.g. if the load of the chain is 1k TPS, can we also read 1k data from the archive?

I haven't load-tested read performance yet. Write performance for ethereum block is in the description above. I'll need to play around with tx generator later to test performance for bigger blocks.

File: monad-archive/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    let s3_archive_writer = S3ArchiveWriter::new(s3_archive).await?;
+
+    let mut latest_processed_block = (s3_archive_writer.get_latest().await).unwrap_or_default();
+
+    info!("Latest processed block is : {}", latest_processed_block);
+
+    let concurrency = args.max_concurrency_level;
+
+    // Check for new blocks every 100 ms
+    // Issue requests to triedb, poll data and push to relevant tables
```
Comment (philipglazman) [inline monad-archive/src/main.rs:45]: I added a stream type called blockwatcher that can be useful here. Maybe it needs its own crate? https://github.com/monad-crypto/monad-bft/blob/master/monad-rpc/src/block_watcher.rs

Comment (tongzhi1998) [reply monad-archive/src/main.rs:45]: Yeah, I would just replace the triedb part with this I guess? I was copying triedb.rs into the monad-archive crate for similar reason. I think we should discuss about how to consolidate the code.

**Pull request closed without merging**